app/models/reactions.js

```diff
@@ -0,0 +1,77 @@
+const _ = require('lodash')
+const mongoose = require('mongoose')
+const Schema = require('mongoose').Schema
+const errors = require('@/errors')
+
+const app = require('@/app')
+
+var Model = module.exports = Schema({
+  kind: {
+    type: String,
+    required: true,
+    enum: ['like', 'recommendation', 'star']
+  },
+
+  comment: {
+    type: Schema.Types.ObjectId,
+    required: true,
+    ref: 'comments'
+  },
+
+  user: {
+    type: Schema.Types.ObjectId,
+    required: true,
+    ref: 'users'
+  },
+
+  active: {
+    type: Boolean,
+    default: true
+  },
+
+  slug: {
+    type: String,
+  }
+})
+
+Model.pre('save', async function(){
+  // Validate if reaction is equal
+  let slug = `${this.kind}:${this.comment._id}:${this.user._id}`
+  if(this.isNew) {
+    let equalReaction = await this.constructor.findOne({ slug: slug })
+    if(equalReaction) throw new errors.BadRequest(`You cannot react 2 identical times in the same comment`)
+    this.slug = slug
+  }
+
+  await validateRules(this)
+})
+
+async function validateRules(reaction){
+  if(this.kind == 'recommendation') {
+    // TODO valida se usuario ja fez materia com aquele professor
+  }
+}
+
+Model.post('save', async function(){
+  await computeReactions(this)
+})
+
+Model.post('remove', async function(){
+  await computeReactions(this)
+})
+
+async function computeReactions(doc) {
+
+  const Comment = app.models.comments
+
+  let comments = await Comment.find({ active: true })
+
+  await Promise.all(comments.map(async function(a)  {
```

Comment:
Mesma análise de antes, talvez tenha uma forma melhor de se fazer isso

Reply:
Na verdade uma forma melhor aqui, é se for uma reaction to tipo 'like', você só atualiza os likes...

```js
await Comment.findOneAndUpdate({ _id: doc.comment._id || doc.comment },
 { `reactionsCount.${doc.kind}`: { $inc: 1 } })
```
ou

```js
const commentId = doc.comment._id || doc.comment
await Comment.findOneAndUpdate({ _id: commentId },
 { `reactionsCount.${doc.kind}`: await doc.constructor.count({ comment: commentId, kind: doc.kind }) })
```

Reply:
Corretissimo 2